contracts/koh/contracts/KingAutomaton.sol

```diff
@@ -1,142 +1,139 @@
 pragma solidity ^0.6.2;
 
-contract KingAutomaton {
-  //////////////////////////////////////////////////////////////////////////////////////////////////
-  // Initialization
-  //////////////////////////////////////////////////////////////////////////////////////////////////
-
-  // TODO(kari): write withdraw() function to replace .transfer
-
-  constructor(uint256 nSlots, uint8 minDifficultyBits, uint256 predefinedMask, uint256 initialDailySupply,
-    int256 approval_pct, int256 contest_pct, uint256 treasury_limit_pct) public {
-    numSlots = nSlots;
-    initMining(nSlots, minDifficultyBits, predefinedMask, initialDailySupply);
-    initNames();
-    initTreasury();
-
-    require(approval_pct > contest_pct, "Approval percentage must be bigger than contest percentage!");
-    approvalPercentage = approval_pct;
-    contestPercentage = contest_pct;
-    treasuryLimitPercentage = treasury_limit_pct;
-    // Check if we're on a testnet (We will not using predefined mask when going live)
-    if (predefinedMask != 0) {
-      // If so, fund the owner for debugging purposes.
-      debugging = true;
-      mint(msg.sender, 1000000 ether);
-      balances[treasuryAddress] = 1000000 ether;
-    }
+library Helpers {
+  // Returns the Ethereum address corresponding to the input public key.
+  function getAddressFromPubKey(bytes32 pubkeyX, bytes32 pubkeyY) public pure returns (uint256) {
+    return uint256(keccak256(abi.encodePacked(pubkeyX, pubkeyY))) & 0x00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
   }
 
-  //////////////////////////////////////////////////////////////////////////////////////////////////
-  // ERC20 Token
-  //////////////////////////////////////////////////////////////////////////////////////////////////
+  // Verifies that signature of a message matches the given public key.
+  function verifySignature(bytes32 pubkeyX, bytes32 pubkeyY, bytes32 hash, uint8 v, bytes32 r, bytes32 s) public pure returns (bool) {
+    uint256 addr = getAddressFromPubKey(pubkeyX, pubkeyY);
+    address addr_r = ecrecover(hash, v, r, s);
+    return addr == uint256(addr_r);
+  }
+}
 
-  // Special purpose community managed addresses.
-  address treasuryAddress = address(1);
-  address DEXAddress = address(2);
+library DEX {
```

Comment:
Helpers, DEX and probably Miner too could go back to the contract due to much higher gas usage